File: days/day11.py
#!/usr/bin/env python
from collections import defaultdict


def part1(data, N=25):
    stones = {int(s):1 for s in data.split()}

    for _ in range(N):
        state = defaultdict(int)
        for x,n in stones.items():
            if x == 0:
                state[1] += n
            else:
                s = str(x)
                t = len(s)
                if t % 2 == 0:
                    a,b = map(int, [s[:t//2], s[t//2:]])
                    state[a] += n
                    state[b] += n
                else:
                    state[x * 2024] += n
        stones = state
            
    ans = sum(stones.values())
    return ans


def part2(data):
    return part1(data, N=75)


data = '''
125 17
'''
assert part1(data) == 55312


data = open('day11.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)

File: days/day11.jl
#!/usr/bin/env julia


function blink(x::Int)
    if x == 0
        [1]
    elseif iseven(ndigits(x))
        divrem(x, 10 ^ (ndigits(x) ÷ 2))
    else
        [x * 2024]
    end
end


function solve(data, N)
    data = parse.(Int, split(data))
    stones = Dict(x=>1 for x in data)
    for _ in 1:N
        state = Dict{Int,Int}()
        for (x, n) in stones, s in blink(x)
            state[s] = get(state, s, 0) + n
        end
        stones = state
    end
    return sum(values(stones))
end


part1(data) = solve(data, 25)
part2(data) = solve(data, 75)


data = """
125 17
"""
@assert part1(data) == 55312


data = readchomp("day11.in")
println("part1: ", part1(data))
println("part2: ", part2(data))
